### PR TITLE
Add GPU based sample laser sensor to simulate Livox mid360 sensor

### PR DESCRIPTION
**Pull request by georgflick**
This PR adds the following:
- commits from the main gazebo repo: [https://github.com/gazebosim/gazebo-classic](https://github.com/gazebosim/gazebo-classic) to keep our devel-enway up-to-date
- commits from Jacobs PR from the cube #3205

My changes:
  - Added a new protobuf type `laserscan_angles`, which extends the type `laserscan` with the `azimuth` and `zenith` angles
  - To use the sample_sensor, you need to set the parameters:
    - `sample_sensor` to true
    - `sample_size` to define the amount of "samples", which are published using the angles in the csv files for each step
    - `sample_csv_file` to a csv file with the angles for the sample sensor
    - to reduce the gpu load while simulating multiple sensors, the texture image was changed from float32 RGB values to float16 R values. Using only one float16 value per pixel, it is not possible to simulate the intensity. The intensity is then set to a fixed value.

How-To-Test:
- Test together with PR in main repo

### CODE REVIEW

File: gazebo/rendering/GpuLaser.cc [hunk 1]
```diff
@@ -427,59 +439,67 @@ GpuLaser::DataIter GpuLaser::LaserDataBegin() const
 {
   const unsigned int index = 0;
   // Data stuffed into three floats (RGB)
-  const unsigned int skip = 3;
+  unsigned int skip = 3;
```
Comment (Roboterbastler) [inline gazebo/rendering/GpuLaser.cc]: This can be constexpr right?

File: gazebo/rendering/GpuLaser.cc [hunk 2]
```diff
@@ -427,59 +439,67 @@ GpuLaser::DataIter GpuLaser::LaserDataBegin() const
...
 GpuLaser::DataIter GpuLaser::LaserDataEnd() const
 {
-  const unsigned int index = this->dataPtr->h2nd * this->dataPtr->w2nd;
+
+  unsigned int index;
+  index = this->dataPtr->verticalRangeCount * this->dataPtr->horizontalRangeCount;
 
   // Data stuffed into three floats (RGB)
-  const unsigned int skip = 3;
+  unsigned int skip = 3;
```
Comment (Roboterbastler) [inline gazebo/rendering/GpuLaser.cc]: Similar here, both can be const/constexpr, no need to assign `index` in next line.

File: gazebo/rendering/GpuLaser.cc (hunk 1 above)
Comment (Roboterbastler) [reply gazebo/rendering/GpuLaser.cc]: Actually the other const variables could also be made constexpr here

File: gazebo/sensors/GpuRaySensor.cc [hunk 3]
```diff
@@ -698,73 +787,152 @@ bool GpuRaySensor::UpdateImpl(const bool /*_force*/)
...
+          scan->add_azimuth(ignition::math::NAN_F);
+          scan->add_zenith(ignition::math::NAN_F);
+        }
+      }
 
-    range = ignition::math::isnan(range) ? this->dataPtr->rangeMax : range;
-    scan->set_ranges(i, range);
-    scan->set_intensities(i, intensity);
-  }
+      // ToDo Georg replace!
```
Comment (Roboterbastler) [inline gazebo/sensors/GpuRaySensor.cc]: Left over todo comment?

File: gazebo/rendering/GpuLaser.cc [hunk 4]
```diff
@@ -427,59 +439,67 @@ GpuLaser::DataIter GpuLaser::LaserDataBegin() const
...
-      intenOffset, this->dataPtr->w2nd);
+  return {index, this->dataPtr->laserBuffer.data(), skip, rangeOffset,
+          intenOffset, this->dataPtr->horizontalRangeCount};
+}
+
+//////////////////////////////////////////////////
+std::vector<float>* GpuLaser::LaserData() const
+{
+  return &this->dataPtr->laserBuffer;
 }
```
Comment (Roboterbastler) [inline gazebo/rendering/GpuLaser.cc]: Return const reference to vector instead of pointer

File: gazebo/sensors/GpuRaySensor.cc [hunk 5]
```diff
@@ -698,73 +787,152 @@ bool GpuRaySensor::UpdateImpl(const bool /*_force*/)
...
+      std::vector<float>* data = this->dataPtr->laserCam->LaserData();
+      constexpr double fixed_intensity = 1.0;
```
Comment (Roboterbastler) [inline gazebo/sensors/GpuRaySensor.cc]: Maybe it would make sense to make this a configurable parameter?

**Review by Roboterbastler — CHANGES_REQUESTED** [review]
Wouldn't it make sense to make it a general option to disable the intensity configuration? The motivation is purely performance related right?
Maybe it would also be beneficial for other (non-sample) sensors if this could be disabled and we save a lot of computation (time). I think we don't use the intensities much, if we even use them anywhere at all...

**Comment by Roboterbastler** [comment]
Should we drop the existing commits concerning the cubemap implementation from `devel-enway`? Maybe it would be much cleaner if the commits in this PR replace those.

File: gazebo/sensors/GpuRaySensor.cc [hunk 6]
```diff
@@ -804,6 +972,69 @@ double GpuRaySensor::NextRequiredTimestamp() const
...
+          }
+          datas.push_back(data);
+      }
+      gzdbg << "data size:" << datas.size() << "\n";
+      return true;
+  } else {
+      gzerr << "cannot read csv file!" << file_name << "\n";
+  }
+  return false;
+}
```
Comment (Roboterbastler) [inline gazebo/sensors/GpuRaySensor.cc]: Seems to be copied from livox plugin, maybe add notice to license on top to be on the safe side?

Comment (georgflick) [reply gazebo/sensors/GpuRaySensor.cc]: Ah right, the copy of the function was more or less for testing. I replaced it now with my own function to load the csv file.

File: gazebo/sensors/GpuRaySensor.cc (hunk 5 above)
Comment (georgflick) [reply gazebo/sensors/GpuRaySensor.cc]: Done, you can now set the parameter `fixed_intensity` and if the parameter `compute_intensity` is set to `false`, the intensity is set to `fixed_intensity`.

File: gazebo/rendering/GpuLaser.cc (hunk 4 above)
Comment (georgflick) [reply gazebo/rendering/GpuLaser.cc]: Done

File: gazebo/sensors/GpuRaySensor.cc (hunk 3 above)
Comment (georgflick) [reply gazebo/sensors/GpuRaySensor.cc]: Removed

**Comment by georgflick** [comment]
> Wouldn't it make sense to make it a general option to disable the intensity configuration? The motivation is purely performance related right? Maybe it would also be beneficial for other (non-sample) sensors if this could be disabled and we save a lot of computation (time). I think we don't use the intensities much, if we even use them anywhere at all...

True, I added a parameter `compute_intensity` , if that is off the texture is only `R_FLOAT16` and intensity is set to the parameter `fixed_intensity`.

**Review by Roboterbastler — APPROVED** [review]
I tested it again and looks good!

**Comment by georgflick** [comment]
Because of the merging conflicts, from updating the commits to the newest version 11.2.0 from the upstream gazebo-classic repo, I will push the gazebo11 PR branch manually to `devel-enway` and merge the PR afterwards.

**Pull request merged**